source/daplink/drag-n-drop/virtual_fs.c

```diff
@@ -295,8 +295,19 @@ void vfs_init(const vfs_filename_t drive_name, uint32_t disk_size)
     init_complete = false;
     // Initialize MBR
     memcpy(&mbr, &mbr_tmpl, sizeof(mbr_t));
-    mbr.total_logical_sectors = ((disk_size + KB(64)) / mbr.bytes_per_sector);
-    mbr.logical_sectors_per_fat = (3 * (((mbr.total_logical_sectors / mbr.sectors_per_cluster) + 1023) / 1024));
+    if (((disk_size + KB(64)) / mbr.bytes_per_sector) >= 32768) {
+        mbr.total_logical_sectors = 0;
+        mbr.big_sectors_on_drive = ((disk_size + KB(64)) / mbr.bytes_per_sector);
+        util_assert((mbr.big_sectors_on_drive / 4096) <= 64);
+        mbr.sectors_per_cluster = 0x10;
+        while ((mbr.big_sectors_on_drive / mbr.sectors_per_cluster) >= 4096) {
+            mbr.sectors_per_cluster *= 2;
+        }
+        mbr.logical_sectors_per_fat = (3 * (((mbr.big_sectors_on_drive / mbr.sectors_per_cluster) + 1023) / 1024));
+    } else {
+        mbr.total_logical_sectors = ((disk_size + KB(64)) / mbr.bytes_per_sector);
+        mbr.logical_sectors_per_fat = (3 * (((mbr.total_logical_sectors / mbr.sectors_per_cluster) + 1023) / 1024));
+    }
```

Comment:
Is this to add support for fat16 in addition to fat 12?  If so everything should probably be switched over to fat16.  I have a patch for this somewhere and can see about bringing it in.